### PR TITLE
[FSDP] Fix `clip_grad_norm_()` for low prec grads

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90028
* #90027

For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1171,12 +1171,21 @@ def clip_grad_norm_(
...
 
-        .. note:: This is analogous to ``torch.nn.utils.clip_grad_norm_`` but
-            handles the partitioning and multiple devices per rank under the
-            hood. The default torch util is not applicable here, because each
-            rank only has a partial view of all the grads in the model, so
-            calling it for FSDP models would lead to different scaling being
-            applied per subset of model parameters.
+        .. note:: If every FSDP instance uses ``NO_SHARD``, meaning that no
+            gradients are sharded across ranks, then you may directly use
+            :func:`torch.nn.utils.clip_grad_norm_`.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1176]: Can we warn explicitly about this?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1171,12 +1171,21 @@ def clip_grad_norm_(
...
+            one other than ``NO_SHARD``), then you should use this method
+            instead of :func:`torch.nn.utils.clip_grad_norm_` since this method
+            handles the fact that gradients are sharded across ranks.
+
+        .. note:: The total norm returned will have the "largest" dtype across
+            all parameters/gradients as defined by PyTorch's type promotion
+            semantics. For example, if *all* parameters/gradients use a low
+            precision dtype, then the returned dtype will be that low precision
+            dtype, but if there exists at least one parameter/gradient using
+            FP32, then the returned dtype will be FP32.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1188]: nit: "then the returned norm will have dtype fp32"

**Review by rohan-varma — APPROVED** [review]
Great catch! Do we know why this was not picked up by unittests earlier?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1176]: I have it so that if all instances use `NO_SHARD`, then this method returns `torch.nn.utils.clip_grad_norm_()`, so it is equivalent.

**Comment by awgu** [comment]
> Great catch! Do we know why this was not picked up by unittests earlier?

We did not not test `keep_low_precision_grads=True` previously :/
